tsl/src/continuous_aggs/materialize.c

```diff
@@ -330,404 +413,368 @@ build_merge_update_clause(List *column_names)
-							const int32 chunk_id)
+create_materialization_plan(MaterializationContext *context, MaterializationPlanType plan_type)
 {
-	StringInfo chunk_condition = makeStringInfo();
-	uint64 rows_processed = 0;
+	MaterializationPlan *materialization = get_materialization_plan(plan_type);
 
-	/* MERGE statement is available starting on PG15 and we'll support it only in the new format of
-	 * CAggs and for non-compressed hypertables */
-	if (ts_guc_enable_merge_on_cagg_refresh && PG_VERSION_NUM >= 150000 &&
-		ContinuousAggIsFinalized(cagg) && !TS_HYPERTABLE_HAS_COMPRESSION_ENABLED(mat_ht))
-	{
-		rows_processed = spi_merge_materializations(mat_ht,
-													cagg,
-													partial_view,
-													materialization_table,
-													time_column_name,
-													invalidation_range);
-	}
-	else
+	if (materialization->plan == NULL)
 	{
-		/*
-		 * chunk_id is valid if the materializaion update should be done only on the given chunk.
-		 * This is used currently for refresh on chunk drop only. In other cases, manual
-		 * call to refresh_continuous_aggregate or call from a refresh policy, chunk_id is
-		 * not provided, i.e., invalid.
-		 */
-		if (chunk_id != INVALID_CHUNK_ID)
-			appendStringInfo(chunk_condition, "AND chunk_id = %d", chunk_id);
-
-		rows_processed += spi_delete_materializations(materialization_table,
-													  time_column_name,
-													  invalidation_range,
-													  chunk_condition->data);
-		rows_processed += spi_insert_materializations(mat_ht,
-													  cagg,
-													  partial_view,
-													  materialization_table,
-													  time_column_name,
-													  invalidation_range,
-													  chunk_condition->data);
+		char *query = create_materialization_plan_query(context, plan_type);
+		Oid types[] = { context->materialization_range.type, context->materialization_range.type };
+
+		materialization->plan = SPI_prepare(query, 2, types);
+		if (materialization->plan == NULL)
+			elog(ERROR, "%s: SPI_prepare failed: %s", __func__, query);
+
+		SPI_keepplan(materialization->plan);
 	}
+}
 
-	/* Get the max(time_dimension) of the materialized data */
-	if (rows_processed > 0)
+static void
+create_materialization_plans(MaterializationContext *context, MaterializationOperation type)
+{
+	switch (type)
 	{
-		spi_update_watermark(mat_ht,
-							 materialization_table,
-							 time_column_name,
-							 invalidation_range,
-							 chunk_condition->data);
+		case OPERATION_INSERT_DELETE:
+			create_materialization_plan(context, PLAN_TYPE_INSERT);
+			create_materialization_plan(context, PLAN_TYPE_DELETE);
+			break;
+		case OPERATION_MERGE_DELETE:
+			create_materialization_plan(context, PLAN_TYPE_INSERT);
+			create_materialization_plan(context, PLAN_TYPE_EXISTS);
+			create_materialization_plan(context, PLAN_TYPE_MERGE);
+			create_materialization_plan(context, PLAN_TYPE_MERGE_DELETE);
+			break;
+		default:
+			pg_unreachable();
```

Comment:
This explicit `create_materialization_plans` requires this function to be called before each usage, which adds to maintenance since you for each code change need to ensure that:

- This function is called before each use of `get_materialization_plan`
- The correct plan type is provided to each call of `create_materialization_plan`.

In contrast, if you move the "prepare code" to `get_materialization_plan` and prepare the plan if it is not already prepared you are always guaranteed that the plan is prepared regardless of how and when the `get_materialization_plan` is called. This would completely eliminate the need for this function and also avoid mistakes in future code where you miss one of the above.

Reply:
Done in that way and removed the useless code!